File: web/src/helpers/storage/partition.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { apiModel } from "~/api/storage";
import { copyApiModel, findDevice, buildPartition } from "~/helpers/storage/api-model";
import { isUsed } from "~/helpers/storage/search";
import { data } from "~/types/storage";

type Partitionable = apiModel.Drive | apiModel.MdRaid;

function indexByName(device: Partitionable, name: string): number {
  return (device.partitions || []).findIndex((p) => p.name && p.name === name);
}

function indexByPath(device: Partitionable, path: string): number {
  return (device.partitions || []).findIndex((p) => p.mountPath === path);
}

/**
 * Adds a new partition.
 *
 * If a partition already exists in the model (e.g., as effect of using the custom policy), then
 * the partition is replaced.
 * */
function addPartition(
  apiModel: apiModel.Config,
  list: "drives" | "mdRaids",
  listIndex: number | string,
  data: data.Partition,
): apiModel.Config {
  apiModel = copyApiModel(apiModel);
  const device = findDevice(apiModel, list, listIndex);

  if (device === undefined) return apiModel;

  // Reset the spacePolicy to the default value if the device goes from unused to used
  if (!isUsed(apiModel, list, listIndex) && device.spacePolicy === "keep")
    device.spacePolicy = null;

  const partition = buildPartition(data);
  const index = indexByName(device, partition.name);

  if (index === -1) device.partitions.push(partition);
  else device.partitions[index] = partition;

  return apiModel;
}

function editPartition(
  apiModel: apiModel.Config,
  list: "drives" | "mdRaids",
  listIndex: number | string,
  mountPath: string,
  data: data.Partition,
): apiModel.Config {
  apiModel = copyApiModel(apiModel);
  const device = findDevice(apiModel, list, listIndex);

  if (device === undefined) return apiModel;

  const index = indexByPath(device, mountPath);
  if (index === -1) return apiModel;

  const oldPartition = device.partitions[index];
  const newPartition = { ...oldPartition, ...buildPartition(data) };
  device.partitions.splice(index, 1, newPartition);

  return apiModel;
}

function deletePartition(
  apiModel: apiModel.Config,
  list: "drives" | "mdRaids",
  listIndex: number | string,
  mountPath: string,
): apiModel.Config {
  apiModel = copyApiModel(apiModel);
  const device = findDevice(apiModel, list, listIndex);

  if (device === undefined) return apiModel;

  const index = indexByPath(device, mountPath);
  device.partitions.splice(index, 1);

  // Do not delete anything if the device is not really used
  if (!isUsed(apiModel, list, listIndex)) {
    device.spacePolicy = "keep";
  }

  return apiModel;
}

export { addPartition, editPartition, deletePartition };

File: service/lib/agama/storage/devicegraph_conversions/to_json_conversions/partition.rb
# frozen_string_literal: true

# Copyright (c) [2025-2026] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/devicegraph_conversions/to_json_conversions/section"
require "y2storage/partition_id"

module Agama
  module Storage
    module DevicegraphConversions
      module ToJSONConversions
        # Section with properties for partitions.
        class Partition < Section
          # @see Section.apply?
          def self.apply?(storage_device)
            storage_device.is?(:partition)
          end

        private

          # @see Section#conversions
          def conversions
            {
              efi: partition_efi,
              id:  partition_id
            }
          end

          # Whether it is a (valid) EFI System partition
          #
          # @return [Boolean]
          def partition_efi
            storage_device.efi_system?
          end

          # Partition Id
          #
          # @return [String, nil] nil if the id is unknown
          def partition_id
            return if storage_device.id == Y2Storage::PartitionId::UNKNOWN

            storage_device.id.to_s
          end
        end
      end
    end
  end
end
